pinot-core/src/main/java/com/linkedin/pinot/core/common/DataFetcher.java

```diff
@@ -107,6 +119,42 @@ public void fetchSingleDictIds(String column, int[] inDocIds, int inStartPos, in
   }
 
   /**
+   * Fetch the dictionary Ids for a multi value column.
+   *
+   * @param column column name.
+   * @param inDocIds document Id array.
+   * @param inStartPos input start position.
+   * @param length input length.
+   * @param outDictIdsArray dictionary Id array array buffer.
+   * @param outStartPos output start position.
+   */
+  public void fetchMultiValueDictIds(String column, int[] inDocIds, int inStartPos, int length, int[][] outDictIdsArray,
+      int outStartPos) {
+    BlockMultiValIterator iterator = (BlockMultiValIterator) getBlockValSetForColumn(column).iterator();
+    int maxNumberOfEntries = getMaxNumberOfEntriesForColumn(column);
+    int[] dictIdArray = new int[maxNumberOfEntries];
+    for (int i = inStartPos; i < inStartPos + length; ++i) {
+      iterator.skipTo(inDocIds[i]);
+      int dictIdLength = iterator.nextIntVal(dictIdArray);
+      outDictIdsArray[outStartPos++] = Arrays.copyOfRange(dictIdArray, 0, dictIdLength);
+
+      if (dictIdLength == 0 ) { 
+        System.out.println("DataFecther.fetchMultiValueDictIds(): dictIdLength = 0,  outDictIdsArray[outStartPos] = " +  outDictIdsArray[outStartPos - 1]);
+      }
+    }
+  }
+
+  /**
+   * For a given multi-value column, trying to get the max number of
+   * entries per row.
+   * @param column
+   * @return max number of entries for a given column.
+   */
+  public int getMaxNumberOfEntriesForColumn(String column) {
```

Comment:
This is used for getting max number of entries for a multi-value column. Then init the temp array. Since that array is cached, this method will only be called once per column.